OSD-20513: Add support for the `--noproxy` directive in `curl`

joshbranham opened the pull request:
<!-- Type of change
Please mark this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement/Feature
- Refactoring
- Tests
- Configuration
- Docs
-->

## What does this PR do? / Related Issues / Jira

This adds an additional `Proxy` config to support passing `--noproxy` to `curl` when probing urls. The values passed will _not_ be proxied through the configured http/https proxy.

I opted to not update the legacy probe based on it being called "legacy" but am open to go down that route if we still have use cases for it.

I tested this with the CLI locally and ran the integration tests as well. Local execution of the `no_proxy` behavior via environment variable:

```
> export http_proxy=squid.corp.redhat.com:3128 https_proxy=squid.corp.redhat.com:3128 no_proxy=google.com
> curl redhat.com -w "%{json}" -Z -s -o /dev/null | jq .proxy_used
1
> curl google.com -w "%{json}" -Z -s -o /dev/null | jq .proxy_used
0
```

On the integration tests, I ran into an aws sdk issue that required bumping the dependencies.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have tested the functionality against gcp / aws, it doesn't cause any regression
- [x] I have added execution results to the PR's readme

## Reviewer's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ] (This needs to be done after technical review) I've run the branch on my local, verified that the functionality is ok

## How to test this PR locally / Special Instructions

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

We do not have the integration test setup with an http/https proxy at this time, but you can modify the integration to pass a `Proxy.NoProxy` slice and validate the tests run as expected. We could invest in including a proxy if we want a more easy test.

## Logs 

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

joshbranham commented:
/label feature

abyrne55 commented on pkg/verifier/aws/entry_point.go:
```diff
@@ -163,6 +163,7 @@ func (a *AwsVerifier) ValidateEgress(vei verifier.ValidateEgressInput) *output.O
 		"HTTP_PROXY":       vei.Proxy.HttpProxy,
 		"HTTPS_PROXY":      vei.Proxy.HttpsProxy,
 		"CACERT":           base64.StdEncoding.EncodeToString([]byte(vei.Proxy.Cacert)),
+		"NOPROXY":          vei.Proxy.NoProxyAsString(),
```
```suggestion
		"NO_PROXY":          vei.Proxy.NoProxyAsString(),
```
nit: staying consistent with our other proxy-related metavars

abyrne55 commented on pkg/verifier/gcp/entry_point.go:
```diff
@@ -95,6 +95,7 @@ func (g *GcpVerifier) ValidateEgress(vei verifier.ValidateEgressInput) *output.O
 		"HTTP_PROXY":       vei.Proxy.HttpProxy,
 		"HTTPS_PROXY":      vei.Proxy.HttpsProxy,
 		"CACERT":           base64.StdEncoding.EncodeToString([]byte(vei.Proxy.Cacert)),
+		"NOPROXY":          vei.Proxy.NoProxyAsString(),
```
```suggestion
		"NO_PROXY":          vei.Proxy.NoProxyAsString(),
```
nit: staying consistent with our other proxy-related metavars

abyrne55 reviewed (changes requested):
Very nice work overall, @joshbranham, and thank you for cleaning up some smelly code in the CLI. I made a few suggestions re. passing NO_PROXY as an env. var vs. a curl flag

joshbranham commented:
/label tide/merge-method-squash

joshbranham commented:
@abyrne55 addressed feedback and ran another set of tests. See below for the curl behavior as tested on it's own as well.

```
> export http_proxy=squid.corp.redhat.com:3128 https_proxy=squid.corp.redhat.com:3128 no_proxy=google.com
> curl redhat.com -w "%{json}" -Z -s -o /dev/null | jq .proxy_used
1
> curl google.com -w "%{json}" -Z -s -o /dev/null | jq .proxy_used
0
```

abyrne55 commented:
Tested this locally using a modified version of examples/aws/terraform/vpc-proxied-explicit. All successful. Good stuff @joshbranham. I'm approving, but before I LGTM, did we not want to put quotes around the NO_PROXY metavar in the userdata template, just in case a space sneaks in there? 

/approve

joshbranham commented:
> Tested this locally using a modified version of examples/aws/terraform/vpc-proxied-explicit. All successful. Good stuff @joshbranham. I'm approving, but before I LGTM, did we not want to put quotes around the NO_PROXY metavar in the userdata template, just in case a space sneaks in there?
> 
> /approve

It shouldn't be possible based on `strings.Join(slice, ",")` but I added quotes on both templates just to be sure 👍

abyrne55 reviewed (approved):
👌💯

/lgtm
